eclass/dist-kernel-utils.eclass

```diff
@@ -26,6 +26,12 @@
 # CONFIG_EFI_ZBOOT enabled. This effects the name of the kernel image on
 # arm64 and riscv. Mainly useful for sys-kernel/gentoo-kernel-bin.
 
+# @ECLASS_VARIABLE: KERNEL_INITRAMFS_GENERATOR
+# @DESCRIPTION:
+# Defines the initramfs generator to be used.
+# By default, dracut is used.
+KERNEL_INITRAMFS_GENERATOR=${KERNEL_INITRAMFS_GENERATOR:=dracut}
```

Comment:
Patches to eclasses should eventually be e-mailed to the gentoo-dev mailing list, but we can discus this here for now.

I like this idea, I'm just wondering if we need some extra abstraction here. For example, if someone were to write an ebuild for mkinitcpio and would want to use it for the distribution kernels, this would not work. For this example `--kernel-image` would be `--kernelimage`, `--kver` would be `--kernelversion`.

Here's my idea:
- The initrd generator must know at least three things:
  -  the path of the kernel image (for uki support)
  - the kernel version (for including the right modules)
  - where to put the resulting image so the eclass can find it and later on install it
- The eclass calls an abstraction tool similar to `installkernel` with these three arguments, this basically does a bunch of if/else to map the arguments to what the initrd generator expects them to be.
- Users can expand this abstraction tool to suite their needs (call custom tools, add extra arguments etc). How this should work I haven't thought out completely yet.

I also think we should expand the `dist-kernel_install_kernel` function below, currently it installs the file named `initrd` or `uki.efi` in the same directory as the kernel image. But according to the manual we can actually provide the path to the initrd we want to use as the fourth argument. We should ensure though that this actually works and that the relevant kernel-install plugins respect this argument (both with and without the systemd implementation of kernel-install).

Another potentially simpler option would be to not generate the initrd in the eclass at all, but instead rely on kernel-install to do this for us. This is easier because kernel-install already has an established way of customizing the kernel installation process via plugins, you can add your own and/or override the ones shipped with installed packages. However, if we do this we have to rethink how `dist-kernel_reinstall_initramfs` works, currently it is used in a couple of ebuilds to trigger the regeneration of the initrd. If we remove initrd generation from the eclass then this function would have to call kernel-install instead to trigger this regeneration.

CC'ing @gentoo/dist-kernel

Reply:
Thank you for the reply. I did not intend for that dist-kernel commit to end up here, but the main reason I'm trying to get ugrd into gentoo repos is to use it with dist-kernel.

I was working on that in #gentoo-dist-kernel, I wanted to add some way to specify arguments but was not sure how to fit that into ebuilds. I just settled on adding args to mine to make it act like dracut, but I think however dist-kernel handles initramfs generation should allow for user input, or at least editing in some ebuild that is designed to fill the role of an initramfs generator for dist-kernel.

I think it would be very nice to be able to change how stuff is passed to installkernel, especially to allow writing to /efi for UKIs and alike.

Reply:
> I think it would be very nice to be able to change how stuff is passed to installkernel, especially to allow writing to /efi for UKIs and alike.

This should already be possible by overriding the `/usr/lib/kernel/install.d/90-loaderentry.install` or `/usr/lib/kernel/install.d/90-uki-copy.install` in `/etc/kernel/install.d/` (files with the same name in /etc override the ones in /usr/lib). 

In fact if I understand the manual correctly you should be able to simply set `BOOT_ROOT="/efi"` in your environment to achieve this:
```
$KERNEL_INSTALL_BOOT_ROOT is set for the plugins to the absolute path of the root directory (mount point, usually) of the hierarchy where boot loader entries, kernel images, and associated resources should be placed. This
usually is the path where the XBOOTLDR partition or the ESP (EFI System Partition) are mounted, and also conceptually referred to as $BOOT. Can be overridden by setting $BOOT_ROOT (see below).
```

Reply:
> I was working on that in #gentoo-dist-kernel, I wanted to add some way to specify arguments but was not sure how to fit that into ebuilds. I just settled on adding args to mine to make it act like dracut, but I think however dist-kernel handles initramfs generation should allow for user input, or at least editing in some ebuild that is designed to fill the role of an initramfs generator for dist-kernel.

@mgorny What was the original rational for using the `dist-kernel-utils.eclass` to generate the initrd/uki? At least with dracut version `059-r4` and up the `kernel-install` plugin can handle generating both an initrd and uki properly. It seems to me we can simplify the eclass a lot, and at the same time make it possible for users to customize the initrd generation and installation process, by dropping all the initrd related code and simply have `installkernel` (which is already called by `dist-kernel_install_kernel` and therefore also by `dist-kernel_reinstall_initramfs` anyway)  take care of this via the established plugin system.

Reply:
> > I think it would be very nice to be able to change how stuff is passed to installkernel, especially to allow writing to /efi for UKIs and alike.
> 
> This should already be possible by overriding the `/usr/lib/kernel/install.d/90-loaderentry.install` or `/usr/lib/kernel/install.d/90-uki-copy.install` in `/etc/kernel/install.d/` (files with the same name in /etc override the ones in /usr/lib).
> 
> In fact if I understand the manual correctly you should be able to simply set `BOOT_ROOT="/efi"` in your environment to achieve this:
> 
> ```
> $KERNEL_INSTALL_BOOT_ROOT is set for the plugins to the absolute path of the root directory (mount point, usually) of the hierarchy where boot loader entries, kernel images, and associated resources should be placed. This
> usually is the path where the XBOOTLDR partition or the ESP (EFI System Partition) are mounted, and also conceptually referred to as $BOOT. Can be overridden by setting $BOOT_ROOT (see below).
> ```
Dist-kernel uses debian's installkernel right? I was thinking there could just be some portage variable that sets the location, I do not have `/usr/lib/kernel` on my machine.

Reply:
> Dist-kernel uses debian's installkernel right? I was thinking there could just be some portage variable that sets the location, I do not have /usr/lib/kernel on my machine.

It uses installkernel which is either provided by installkernel-gentoo (our fork of the original debian tool) or installkernel-systemd. I use the systemd version, but if the installkernel-gentoo version doesn't use this plugin system then I guess this answers my question about why the eclass generates the initrd in the first place.

If that's the case then maybe it is worth to either:
- work on making installkernel-gentoo generate the initrd (so the eclass code can be dropped). User customization of the used initrd would then have to be implemented there, instead of the eclass, which to me makes more sense. The systend implementation of kernel-install already allows for arbitrary customization, and can already generate an initrd or uki with dracut.
- try to decouple systemd's kernel-install from systemd-boot, and use this as the only implementation of installkernel. I think this should be doable since the systemd build system already provides a build option for kernel-install separately from the one for systemd-boot, the question then is if it will actually build in a configuration where kernel-install is enabled but systemd-boot is disabled.

Reply:
@desultory I opened this PR which works towards support of non-dracut initrd generators: https://github.com/gentoo/gentoo/pull/33727

Could you give this a try and let me know if it accomplishes what you want to do?

You'll need to switch out `installkernel-gentoo` for `installkernel-systemd` and maybe write a `kernel-install` plugin for your initrd generator and possibly a plugin for your bootloader as well if it doesn't exist. This is relatively simple, you can use `/usr/lib/kernel/install.d/50-dracut.install` as an example.

Reply:
Just to be sure I'm understanding this correctly, but is installkernel-systemd likely going to be the new standard? I have never used it and I'm not sure how coupled it is to systemd-boot etc. 

If I'm reading correctly, both systems pull some modules from /etc/kernel/ but from different directories. Maybe the installkernel-gentoo script could be updated to use the same modules as the sytstemd variant?

Reply:
> Just to be sure I'm understanding this correctly, but is installkernel-systemd likely going to be the new standard?

No, installkernel-gentoo will remain the first option , I'm just trying to add some flexibility and this is easiest to do in installkernel-systemd because of its plugin system. Maybe in the future, if everything is shipped with a proper kernel-install plugin, we can reconsider this.

> I have never used it and I'm not sure how coupled it is to systemd-boot etc.

My PR uncouples this. There is no reason why you could not use `kernel-install` with some other boot loader AFAIK. All you need is a proper plugin in `/etc/kernel/install.d/` or `/usr/lib/kernel/install.d`. After my PR is merged one could install `sys-apps/systemd-utils[kernel-install]` and this should allow you to use systemd's fancy `kernel-install` independently of systemd and systemd-boot. You might need to set `layout=other` in `/etc/kernel/install.conf` to ensure that the default plugin scripts for the bls and uki layout are not used.

> If I'm reading correctly, both systems pull some modules from /etc/kernel/ but from different directories. Maybe the installkernel-gentoo script could be updated to use the same modules as the sytstemd variant?

Something like this should be possible, but this is more complicated so I'm focussing on installkernel-systemd first where it is easy to allow some customizability by simply moving some stuff around.

Reply:
Just opened a PR to move initrd/uki generation into installkernel-gentoo, which should eventually allow the initrd generation code to be removed from the eclasses altogether. https://github.com/projg2/installkernel-gentoo/pull/10

Reply:
@desultory you should now be able to use arbitrary initrd/uki generators with both installkernel-gentoo and installkernel-systemd.

For installkernel-gentoo an initrd plugin should put a file `initrd` in the same directory as the kernel image (see `/etc/kernel/preinst.d/50-dracut.install` for an example)
For installkernel-systemd an initrd plugin should put a file `initrd` on the `${KERNEL_INSTALL_STAGING_AREA}` (see `/usr/lib/kernel/install.d/50-dracut.install` for an example).

Let me know if this works for you!